### PR TITLE
Make input stack capacity configurable

### PR DESCRIPTION
**Pull request by dapetcu21**
The input stack limit of 16 is very limitative for complex scenes. For example, a complex scene with more than 16 GUIs, each a different panel displaying different information (a situation which cannot be handled very well with a common input manager).

This PR makes the limit configurable through `collection.max_input_stack_entries`.

### CODE REVIEW

**Review by britzl — COMMENTED** [review]
I think it looks good. Your thoughts @JCash ?

File: engine/gameobject/src/gameobject/gameobject.cpp [hunk 1]
```diff
@@ -193,8 +195,7 @@ namespace dmGameObject
         m_WorldTransforms.SetCapacity(max_instances);
         m_WorldTransforms.SetSize(max_instances);
         m_IDToInstance.SetCapacity(dmMath::Max(1U, max_instances/3), max_instances);
-        // TODO: Un-hard-code
-        m_InputFocusStack.SetCapacity(16);
+        m_InputFocusStack.SetCapacity(max_input_stack_entries);
```
Comment (JCash) [inline engine/gameobject/src/gameobject/gameobject.cpp:198]: indentation seems off

File: engine/gameobject/src/gameobject/gameobject.cpp [hunk 2]
```diff
@@ -253,7 +266,7 @@ namespace dmGameObject
 
     Collection* AllocCollection(const char* name, HRegister regist, uint32_t max_instances)
     {
-        Collection* collection = new Collection(0, 0, max_instances);
+        Collection* collection = new Collection(0, 0, max_instances, GetInputStackDefaultCapacity(regist));
```
Comment (JCash) [inline engine/gameobject/src/gameobject/gameobject.cpp:269]: I wonder why the "max_input_stack_entries" doesn't take the same code path as the other configurable "max_instances"?

File: engine/gameobject/src/gameobject/gameobject.cpp (hunk 1 above)
Comment (dapetcu21) [reply engine/gameobject/src/gameobject/gameobject.cpp:198]: Huh? How? Where? Looks ok to me. Are you not using 4 spaces?

File: engine/gameobject/src/gameobject/gameobject.cpp (hunk 2 above)
Comment (dapetcu21) [reply engine/gameobject/src/gameobject/gameobject.cpp:269]: I guess I could make it the same, it's just that that max_instances gets passed around through 3 functions before reaching that constructor and the indirection felt unnecessary in this case. I don't really know why is there so much indirection or what's the better approach here.

File: engine/gameobject/src/gameobject/gameobject.cpp (hunk 1 above)
Comment (JCash) [reply engine/gameobject/src/gameobject/gameobject.cpp:198]: Hmm, I think it's github that does something "clever" with line breaks when displaying the diff that makes it look confusing.

File: engine/gameobject/src/gameobject/gameobject.cpp [hunk 3]
```diff
@@ -237,6 +238,18 @@ namespace dmGameObject
         return regist->m_DefaultCollectionCapacity;
     }
 
+    void SetInputStackDefaultCapacity(HRegister regist, uint32_t capacity)
+    {
+        assert(regist != 0x0);
+        regist->m_DefaultInputStackCapacity = capacity;
+    }
+
+    uint32_t GetInputStackDefaultCapacity(HRegister regist)
```
Comment (JCash) [inline engine/gameobject/src/gameobject/gameobject.cpp:247]: Shouldn't be needed outside of this cpp file?
I recommend adding "static" to this function.

File: engine/gameobject/src/gameobject/gameobject.cpp (hunk 2 above)
Comment (JCash) [reply engine/gameobject/src/gameobject/gameobject.cpp:269]: I can't say I like passing those things around either.
Also, in the future, this library should take a create params struct, like our other systems do. E.g. to dmGameObject::NewRegister(create_params). In order to reduce the amount of overhead for passing/storing/reading the settings.
But, that's for the future. Let's keep this for now.

**Review by JCash — DISMISSED** [review]

File: engine/gameobject/src/gameobject/gameobject.cpp (hunk 3 above)
Comment (dapetcu21) [reply engine/gameobject/src/gameobject/gameobject.cpp:247]: Yeah, you're right. The use site is in the same file. I can add a static.

Comment (dapetcu21) [reply engine/gameobject/src/gameobject/gameobject.cpp]: Done

**Review by JCash — APPROVED** [review]
Looks good

**Pull request merged**